### DISCUSSION TITLE
Render slint in Terminal window

### DISCUSSION THREAD

**Discussion by ji-podhead** [slint-ui/slint#4672] [Ideas] [answered]
I wanted to to create a multiplexer-app for kubernetes so i checked out tview which is nice, but it turns out its just to complicated to create a complex ui (no scrollbar and get-scale/size methods to build your own scrollbar), so i decided to use slint.
I asked phind about how to render slint inside terminal and one time it told me that there was some additional package that could help achieving this.
But i think that can get quite complex, so i decided to ask about that here. 
Is that even possible or to complex?
Thanks in advance!

**Comment by hunger** [comment]
You mean to turn a Slint file into a text-based UI? I am not aware of anything that does that for Slint files.

It could probably be done for _very_ simple cases using layouts, but I see no way to do this in an automated way.

**Reply by ji-podhead** [reply]
hmmm i think its possible somehow.
https://www.redhat.com/sysadmin/music-video-linux-terminal
here they render a yt video in the terminal.
maybe you could place the terminal window in front of the slint video and deactivate the input events for the terminal window, or you redirect the input to the slint window.
idk
but it would be cool if that was possible.
slint makes it easy for me to create layouts, but i really want to have it in the terminal :D :D :D

**Reply by ji-podhead** [reply]
jap ncurses as display driver and then redirect input events could be  a possible solution

**Reply by ji-podhead** [reply]
is the slint community meeting up at hackatons or any other events in berlin this year?
that would be cool

**Reply by ji-podhead** [reply]
but maybe resolution is to low

**Reply by hunger** [reply]
We tend to be at rust meetups in Berlin. Sometimes we join C++ meetups, too. :-)

Yes, you can turn the terminal into a low-res "display" just fine, but that basically blows up each window by a factor of 16 (assuming a character in the terminal consists of 16x16 pixels). That will make your windows huge -- and if you start to shrink that down, you get "blurry text" and similar effects.

I thought you wanted to basically generate TUIs like those in `btm`, `lazygit`, `ranger`, `midnight commander`, `vim`, ... etc. For that you need to map Text to character sequences in the terminal and UI elements like buttons to styled text. So you would need a custom style for text-based rendering. That is probably doable, but will need simple `.slint` files to work and will not automatically make any random slint UI available in the terminal.

**Reply by ji-podhead** [reply]
Hey, thanks for your answer. Unfortunately, I was afraid that the text would be blurry due to the scaling. Of course, having to scale the resolution is not a real solution and could also lead to excessive CPU utilization. Maybe the whole thing is more of a gimmick than really practical, I'll try out what I can do best with it. Maybe I can make the text thick enough from the start so that the blur isn't as noticeable.

I wanted to do this because I'm actually looking for a way to create a controller for K8s.
But if you can integrate the Slint UI as an extension directly into vscode, then that's actually enough for me. I'm looking for a way to automatically retrieve function descriptions and arguments/flags and then trigger them generically. Therefore I think this is best achieved using NodeJS and a Python implementation (inspect class).
I tried using Rust but couldn't really find a solution without hardcoding everything (wrapper classes).
I've never built a vscode extension before, but it seems easy using npm, so I'll try to make that happen for now.
Thanks for your advice.
best regards: ji

**Comment by ogoffart** [comment, marked as answer]
Actually, i've made some experiment before when we were working on the MCU backend, to render everything in a terminal:
https://github.com/slint-ui/slint/commit/5cdc70a027e5460c4528964914dcbbdb79538d70
(But this is a bit old and will not build out of the box.  Might give some ideas though)
